File: packages/widget/test/identity-header.test.ts
// @vitest-environment jsdom

import { describe, expect, it } from 'vitest';

import type { HandledBy } from '@dhaam-ccrm/core';
import type { ChatSession, ChatStatus } from '@dhaam-ccrm/core';

import { createIdentityHeader } from '../src/ui/identity-header.js';

/** A SPECIFIC title — a real store/outlet/merchant name — never paired with a handler's name. */
const FALLBACK = 'Acme Support';
/** A GENERIC title — the widget's own default — IS paired with a handler's name once one is current. */
const GENERIC_FALLBACK = 'Chat with us';

function session(overrides: Partial<Pick<ChatSession, 'status' | 'handledBy'>> = {}): Pick<
  ChatSession,
  'status' | 'handledBy'
> {
  return { status: 'ASSIGNED', ...overrides };
}

const AGENT: HandledBy = { kind: 'AGENT', id: 'agt_1', displayName: 'Ada' };
const BOT: HandledBy = { kind: 'BOT', id: 'bot_1', displayName: 'Assistant' };

describe('mount shape', () => {
  it('is an <h2 id="dh-title">, the id the panel already wires aria-labelledby to', () => {
    const header = createIdentityHeader(FALLBACK);
    expect(header.node.tagName).toBe('H2');
    expect(header.node.id).toBe('dh-title');
  });

  it('starts on the configured title before any update', () => {
    const header = createIdentityHeader(FALLBACK);
    expect(header.node.textContent).toBe(FALLBACK);
  });

  it('exposes a dedicated status live region, separate from the title itself', () => {
    const header = createIdentityHeader(FALLBACK);
    expect(header.liveRegion.getAttribute('role')).toBe('status');
    expect(header.liveRegion.getAttribute('aria-live')).toBe('polite');
    expect(header.liveRegion).not.toBe(header.node);
  });
});

describe('absence — handledBy undefined means "render my own title", not "unhandled"', () => {
  it('falls back to the configured title when session is null', () => {
    const header = createIdentityHeader(FALLBACK);
    header.update(null);
    expect(header.node.textContent).toBe(FALLBACK);
  });

  it('falls back to the configured title when handledBy is simply absent', () => {
    const header = createIdentityHeader(FALLBACK);
    header.update(session({ status: 'WAITING_FOR_AGENT' }));
    expect(header.node.textContent).toBe(FALLBACK);
    header.update(session({ status: 'OPEN' }));
    expect(header.node.textContent).toBe(FALLBACK);
  });
});

describe('a SPECIFIC title (a real store/outlet name) is never paired with who answers', () => {
  // A store-scoped chat is identified by the outlet, not by which staff
  // member happens to be on it — "tse · Mohali" answers a question the
  // customer never asked and buries the one fact that matters. So a
  // specific title stands alone regardless of the session's handler.

  it('stays on the configured title once a human agent is assigned', () => {
    const header = createIdentityHeader(FALLBACK);
    header.update(session({ status: 'ASSIGNED', handledBy: AGENT }));
    expect(header.node.textContent).toBe(FALLBACK);
    // The avatar/CSS hook still reflects the real handler — only the visible
    // title text is unaffected by who it is.
    expect(header.node.getAttribute('data-handled-by')).toBe('AGENT');
  });

  it('stays on the configured title while a bot handles it', () => {
    const header = createIdentityHeader(FALLBACK);
    header.update(session({ status: 'OPEN', handledBy: BOT }));
    expect(header.node.textContent).toBe(FALLBACK);
    expect(header.node.getAttribute('data-handled-by')).toBe('BOT');
  });

  it('every other status agrees — a specific title never changes for a handler', () => {
    const statuses: ChatStatus[] = ['OPEN', 'ASSIGNED', 'CLOSED', 'RESOLVED', 'ON_HOLD'];
    for (const status of statuses) {
      const header = createIdentityHeader(FALLBACK);
      header.update(session({ status, handledBy: AGENT }));
      expect(header.node.textContent).toBe(FALLBACK);
    }
  });
});

describe('a GENERIC title pairs with the current handler’s name', () => {
  it('pairs the human agent with the platform brand once assigned', () => {
    const header = createIdentityHeader(GENERIC_FALLBACK);
    header.update(session({ status: 'ASSIGNED', handledBy: AGENT }));
    expect(header.node.textContent).toBe('Ada · Dhaam Support');
    expect(header.node.getAttribute('data-handled-by')).toBe('AGENT');
  });

  it('pairs the bot the same way while it handles the chat', () => {
    const header = createIdentityHeader(GENERIC_FALLBACK);
    header.update(session({ status: 'OPEN', handledBy: BOT }));
    expect(header.node.textContent).toBe('Assistant · Dhaam Support');
  });

  it('an absent config.title (undefined → "") is generic too', () => {
    const header = createIdentityHeader('');
    header.update(session({ status: 'ASSIGNED', handledBy: AGENT }));
    expect(header.node.textContent).toBe('Ada · Dhaam Support');
  });

  it('shows the bare name once, not paired with itself, when the handler IS the platform brand', () => {
    // The tenant's default bot name ("Assistant") colliding with a
    // placeholder widget title — must not read "Assistant · Dhaam Support"
    // when the bot itself already carries a different name; it only
    // collapses when the two strings are actually identical.
    const header = createIdentityHeader(GENERIC_FALLBACK);
    header.update(session({ status: 'OPEN', handledBy: { kind: 'BOT', id: 'b1', displayName: 'Dhaam Support' } }));
    expect(header.node.textContent).toBe('Dhaam Support');
  });

  it('is case- and whitespace-insensitive about the same-name check', () => {
    // The comparison trims/lowercases before comparing, but still renders the
    // handler's name exactly as the server sent it — trimming the DISPLAYED
    // text is not this guard's job.
    const header = createIdentityHeader(GENERIC_FALLBACK);
    header.update(session({ status: 'OPEN', handledBy: { kind: 'BOT', id: 'b1', displayName: '  dhaam support  ' } }));
    expect(header.node.textContent).toBe('  dhaam support  ');
  });
});

describe('staleness — a reactivated session keeps a name isHandledByCurrent must reject', () => {
  it(
    'renders the configured title, NOT the closing agent\'s name, when a reactivated session ' +
      'reports WAITING_FOR_AGENT with a stale handledBy',
    () => {
      // Exactly core T10's documented case: a session reactivated from
      // CLOSED/RESOLVED keeps its previous assignedAgentId server-side, so
      // handledBy can still name the agent who closed it even though status
      // has already gone back to WAITING_FOR_AGENT. Rendering "Ada" here
      // would tell the customer someone is with them when nobody is.
      const header = createIdentityHeader(GENERIC_FALLBACK);
      header.update(session({ status: 'WAITING_FOR_AGENT', handledBy: AGENT }));

      expect(header.node.textContent).toBe(GENERIC_FALLBACK);
      expect(header.node.textContent).not.toContain('Ada');
      expect(header.node.getAttribute('data-handled-by')).toBe('');
    },
  );
});

describe('the live-region announcement', () => {
  // Exercised against a GENERIC title: that is the one case where the
  // visible label actually moves when a handler joins or leaves, so it is
  // the one case where an announcement has anything to say. A specific
  // title's label never changes for a handler (see above), so it never
  // announces a hand-off either — correctly: the header intentionally does
  // not surface the individual staff member for a store-scoped chat, and
  // that includes the screen-reader channel.

  it('never announces on the very first update — that describes what was already true, not a live change', () => {
    const header = createIdentityHeader(GENERIC_FALLBACK);
    header.update(session({ status: 'ASSIGNED', handledBy: AGENT }));
    expect(header.liveRegion.textContent).toBe('');
  });

  it('announces once an agent joins mid-session', () => {
    const header = createIdentityHeader(GENERIC_FALLBACK);
    header.update(session({ status: 'WAITING_FOR_AGENT' })); // seed: fallback title
    header.update(session({ status: 'ASSIGNED', handledBy: AGENT })); // agent.joined applied

    expect(header.liveRegion.textContent).toBe("You're now chatting with Ada.");
  });

  it('announces the reversion once an agent leaves', () => {
    const header = createIdentityHeader(GENERIC_FALLBACK);
    header.update(session({ status: 'ASSIGNED', handledBy: AGENT }));
    header.liveRegion.textContent = '';

    header.update(session({ status: 'WAITING_FOR_AGENT' })); // agent.left applied
    expect(header.liveRegion.textContent).toBe(`You're now chatting with ${GENERIC_FALLBACK}.`);
  });

  it('does not re-announce when an update reruns with no actual label change', () => {
    // A store selector can re-fire for reasons unrelated to identity (e.g. a
    // new session object with the same handledBy). Repeating the
    // announcement every time would talk over whatever the user is reading —
    // the same discipline message-list.ts's setClosure applies to its own
    // live region.
    const header = createIdentityHeader(GENERIC_FALLBACK);
    header.update(session({ status: 'ASSIGNED', handledBy: AGENT }));
    header.update(session({ status: 'ASSIGNED', handledBy: AGENT }));
    header.liveRegion.textContent = '';

    header.update(session({ status: 'ASSIGNED', handledBy: { ...AGENT } }));
    expect(header.liveRegion.textContent).toBe('');
  });

  it('does not announce a stale handledBy as though it were a real hand-off', () => {
    const header = createIdentityHeader(GENERIC_FALLBACK);
    header.update(session({ status: 'ASSIGNED', handledBy: AGENT }));
    header.liveRegion.textContent = '';

    // Reactivation: handledBy still says Ada, but isHandledByCurrent rejects
    // it — the rendered (and thus announced) label is the fallback title,
    // which differs from 'Ada', so this DOES announce, but with the honest
    // fallback text, never Ada's name.
    header.update(session({ status: 'WAITING_FOR_AGENT', handledBy: AGENT }));
    expect(header.liveRegion.textContent).toBe(`You're now chatting with ${GENERIC_FALLBACK}.`);
  });

  it('never announces at all for a SPECIFIC title — the label never moves for a handler to begin with', () => {
    const header = createIdentityHeader(FALLBACK);
    header.update(session({ status: 'WAITING_FOR_AGENT' }));
    header.update(session({ status: 'ASSIGNED', handledBy: AGENT }));
    expect(header.liveRegion.textContent).toBe('');
    header.update(session({ status: 'WAITING_FOR_AGENT' }));
    expect(header.liveRegion.textContent).toBe('');
  });
});

File: packages/flutter/test/ui/header/identity_header_test.dart
import 'package:dhaam_chat/dhaam_chat.dart';
import 'package:dhaam_chat_flutter/dhaam_chat_flutter.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

/// Reproduces `packages/widget/test/identity-header.test.ts` (12 cases) and
/// the identity half of `identity-header-mount.test.ts`.
///
/// The mount file's other half — "does the widget mount T11's component rather
/// than a second hand-built `<h2>`" — is a DOM single-sourcing question
/// (`querySelectorAll('#dh-title')` returning exactly one node, the panel's
/// `aria-labelledby` still pointing at it). Flutter has no id space and no
/// `aria-labelledby`; the equivalent property is that there is one
/// [IdentityHeader] widget and one [isHandledByCurrent], which is a structural
/// fact rather than something a test can drift away from. What DOES transfer
/// is every case about identity following the session, and those are here.
const String kFallback = 'Acme Support';

const HandledBy agent =
    HandledBy(kind: HandledByKind.agent, id: 'agt_1', displayName: 'Ada');
const HandledBy bot =
    HandledBy(kind: HandledByKind.bot, id: 'bot_1', displayName: 'Assistant');

SessionSnapshot session({
  ChatStatus status = ChatStatus.assigned,
  HandledBy? handledBy,
}) =>
    SessionSnapshot(
      sessionId: 'sess_1',
      status: status,
      mode: ChatMode.human,
      participants: const <ParticipantSnapshot>[],
      createdAt: DateTime.utc(2026, 1, 1),
      handledBy: handledBy,
    );

void main() {
  group('isHandledByCurrent — the one gate', () {
    test('false for an absent handledBy, whatever the status says', () {
      for (final ChatStatus status in ChatStatus.values) {
        expect(isHandledByCurrent(session(status: status)), isFalse);
      }
    });

    test('false for a present handledBy while the session says it is waiting',
        () {
      expect(
        isHandledByCurrent(
          session(status: ChatStatus.waitingForAgent, handledBy: agent),
        ),
        isFalse,
      );
    });

    test('true for a present handledBy on every other status', () {
      for (final ChatStatus status in ChatStatus.values
          .where((ChatStatus s) => s != ChatStatus.waitingForAgent)) {
        expect(
          isHandledByCurrent(session(status: status, handledBy: agent)),
          isTrue,
          reason: '$status',
        );
      }
    });
  });

  group('absence — handledBy absent means "render my own title"', () {
    test('falls back when there is no session at all', () {
      expect(identityLabel(null, kFallback), kFallback);
      expect(identityKind(null), isNull);
    });

    test('falls back when handledBy is simply absent', () {
      expect(
        identityLabel(session(status: ChatStatus.waitingForAgent), kFallback),
        kFallback,
      );
      expect(
        identityLabel(session(status: ChatStatus.open), kFallback),
        kFallback,
      );
    });
  });

  group('presence — a current handledBy names the agent or the bot', () {
    test('shows the human agent once assigned', () {
      final SessionSnapshot s =
          session(status: ChatStatus.assigned, handledBy: agent);
      expect(identityLabel(s, kFallback), 'Ada');
      expect(identityKind(s), HandledByKind.agent);
    });

    test("shows the bot's own name while the bot handles it", () {
      final SessionSnapshot s =
          session(status: ChatStatus.open, handledBy: bot);
      expect(identityLabel(s, kFallback), 'Assistant');
      expect(identityKind(s), HandledByKind.bot);
    });
  });

  group('staleness — a reactivated session keeps a name the gate must reject',
      () {
    test(
        'renders the configured title, NOT the closing agent\'s name, when a '
        'reactivated session reports WAITING_FOR_AGENT with a stale handledBy',
        () {
      // A session reactivated from CLOSED/RESOLVED keeps its previous
      // assigned agent server-side, so handledBy can still name the agent who
      // closed it even though status has gone back to WAITING_FOR_AGENT.
      // Rendering "Ada" here would tell the customer someone is with them
      // when nobody is.
      final SessionSnapshot s =
          session(status: ChatStatus.waitingForAgent, handledBy: agent);
      expect(identityLabel(s, kFallback), kFallback);
      expect(identityLabel(s, kFallback), isNot(contains('Ada')));
      expect(identityKind(s), isNull);
    });
  });

  group('the rendered title', () {
    Future<void> pump(
      WidgetTester tester, {
      SessionSnapshot? session,
      String fallbackTitle = kFallback,
    }) =>
        tester.pumpWidget(
          MaterialApp(
            home: Scaffold(
              body: IdentityHeader(
                session: session,
                fallbackTitle: fallbackTitle,
              ),
            ),
          ),
        );

    testWidgets('starts on the configured title before any session exists',
        (WidgetTester tester) async {
      await pump(tester);
      expect(find.text(kFallback), findsOneWidget);
    });

    testWidgets('names the agent once one is handling the chat',
        (WidgetTester tester) async {
      await pump(tester);
      await pump(
        tester,
        session: session(status: ChatStatus.assigned, handledBy: agent),
      );
      expect(find.text('Ada'), findsOneWidget);
      expect(find.text(kFallback), findsNothing);
    });

    testWidgets('falls back again when the agent leaves',
        (WidgetTester tester) async {
      await pump(
        tester,
        session: session(status: ChatStatus.assigned, handledBy: agent),
      );
      await pump(tester, session: session(status: ChatStatus.waitingForAgent));
      expect(find.text(kFallback), findsOneWidget);
    });

    testWidgets(
        'does not narrate a stale agent on a session that went back '
        'to waiting', (WidgetTester tester) async {
      await pump(
        tester,
        session: session(status: ChatStatus.assigned, handledBy: agent),
      );
      await pump(
        tester,
        session: session(status: ChatStatus.waitingForAgent, handledBy: agent),
      );
      expect(find.text(kFallback), findsOneWidget);
      expect(find.text('Ada'), findsNothing);
    });
  });

  group('the live-region announcement', () {
    /// The announcement currently on the semantics tree, or null.
    ///
    /// The Dart counterpart of the reference test reading
    /// `header.liveRegion.textContent`: a node that carries no visible layout
    /// and exists only while there is something to say.
    String? announcement(WidgetTester tester) {
      final Iterable<Semantics> live = tester
          .widgetList<Semantics>(find.byType(Semantics))
          .where((Semantics s) => s.properties.liveRegion ?? false);
      return live.isEmpty ? null : live.first.properties.label;
    }

    Future<void> pump(
      WidgetTester tester, {
      SessionSnapshot? session,
      String fallbackTitle = kFallback,
    }) =>
        tester.pumpWidget(
          MaterialApp(
            home: Scaffold(
              body: IdentityHeader(
                session: session,
                fallbackTitle: fallbackTitle,
              ),
            ),
          ),
        );

    testWidgets(
        'never announces on the very first build — that describes '
        'what was already true, not a live change',
        (WidgetTester tester) async {
      await pump(
        tester,
        session: session(status: ChatStatus.assigned, handledBy: agent),
      );
      expect(find.text('Ada'), findsOneWidget);
      expect(announcement(tester), isNull);
    });

    testWidgets('announces once an agent joins mid-session',
        (WidgetTester tester) async {
      await pump(tester, session: session(status: ChatStatus.waitingForAgent));
      await pump(
        tester,
        session: session(status: ChatStatus.assigned, handledBy: agent),
      );
      expect(announcement(tester), "You're now chatting with Ada.");
    });

    testWidgets('announces the reversion once an agent leaves',
        (WidgetTester tester) async {
      await pump(
        tester,
        session: session(status: ChatStatus.assigned, handledBy: agent),
      );
      await pump(tester, session: session(status: ChatStatus.waitingForAgent));
      expect(announcement(tester), "You're now chatting with $kFallback.");
    });

    testWidgets(
        'does not re-announce when a rebuild carries no actual label '
        'change', (WidgetTester tester) async {
      // A cubit can emit for reasons unrelated to identity — a new session
      // object with the same handledBy. Repeating the announcement every time
      // would talk over whatever the user is reading.
      await pump(tester, session: session(status: ChatStatus.waitingForAgent));
      await pump(
        tester,
        session: session(status: ChatStatus.assigned, handledBy: agent),
      );
      expect(announcement(tester), "You're now chatting with Ada.");

      await pump(
        tester,
        session: session(
          status: ChatStatus.assigned,
          handledBy: const HandledBy(
            kind: HandledByKind.agent,
            id: 'agt_1',
            displayName: 'Ada',
          ),
        ),
      );
      // Still the same sentence, not a second one — the watermark compares
      // the LABEL, so a fresh object naming the same person changes nothing.
      expect(announcement(tester), "You're now chatting with Ada.");
    });

    testWidgets(
        'does not announce a stale handledBy as though it were a real '
        'hand-off', (WidgetTester tester) async {
      await pump(tester, session: session(status: ChatStatus.waitingForAgent));
      await pump(
        tester,
        session: session(status: ChatStatus.assigned, handledBy: agent),
      );

      // Reactivation: handledBy still says Ada, but the gate rejects it — so
      // the rendered (and thus announced) label is the fallback title, which
      // DOES differ from 'Ada'. This announces, but with the honest fallback
      // text, never Ada's name.
      await pump(
        tester,
        session: session(status: ChatStatus.waitingForAgent, handledBy: agent),
      );
      expect(announcement(tester), "You're now chatting with $kFallback.");
      expect(announcement(tester), isNot(contains('Ada')));
    });
  });

  group('a new fallback title — the setFallbackTitle contract', () {
    String? announcement(WidgetTester tester) {
      final Iterable<Semantics> live = tester
          .widgetList<Semantics>(find.byType(Semantics))
          .where((Semantics s) => s.properties.liveRegion ?? false);
      return live.isEmpty ? null : live.first.properties.label;
    }

    Future<void> pump(
      WidgetTester tester, {
      SessionSnapshot? session,
      required String fallbackTitle,
    }) =>
        tester.pumpWidget(
          MaterialApp(
            home: Scaffold(
              body: IdentityHeader(
                session: session,
                fallbackTitle: fallbackTitle,
              ),
            ),
          ),
        );

    testWidgets('repaints when the fallback is what is on screen',
        (WidgetTester tester) async {
      await pump(tester, fallbackTitle: 'Chat');
      expect(find.text('Chat'), findsOneWidget);
      await pump(tester, fallbackTitle: kFallback);
      expect(find.text(kFallback), findsOneWidget);
    });

    testWidgets("no-ops visually while an agent's name is showing",
        (WidgetTester tester) async {
      // An agent's name outranks a configured title; stamping over it would
      // rename the person the customer is talking to.
      final SessionSnapshot live =
          session(status: ChatStatus.assigned, handledBy: agent);
      await pump(tester, session: live, fallbackTitle: 'Chat');
      await pump(tester, session: live, fallbackTitle: kFallback);
      expect(find.text('Ada'), findsOneWidget);
      expect(find.text(kFallback), findsNothing);
    });

    testWidgets('is SILENT — a landed config fetch is not a hand-off',
        (WidgetTester tester) async {
      // Telling a screen-reader user "You're now chatting with Acme Support"
      // because a config fetch landed is a lie about an event that did not
      // happen.
      await pump(tester, fallbackTitle: 'Chat');
      await pump(tester, fallbackTitle: kFallback);
      expect(find.text(kFallback), findsOneWidget);
      expect(announcement(tester), isNull);
    });

    testWidgets(
        'and it clears a previous announcement rather than leaving it '
        'to be re-read', (WidgetTester tester) async {
      await pump(tester, fallbackTitle: kFallback);
      await pump(
        tester,
        session: session(status: ChatStatus.assigned, handledBy: agent),
        fallbackTitle: kFallback,
      );
      expect(announcement(tester), "You're now chatting with Ada.");

      await pump(
        tester,
        session: session(status: ChatStatus.assigned, handledBy: agent),
        fallbackTitle: 'Acme Help',
      );
      expect(announcement(tester), isNull);
      expect(find.text('Ada'), findsOneWidget);
    });

    testWidgets(
        'the watermark still moves, so a later hand-off is measured '
        'against what is displayed', (WidgetTester tester) async {
      await pump(tester, fallbackTitle: 'Chat');
      await pump(tester, fallbackTitle: kFallback);
      expect(announcement(tester), isNull);

      await pump(
        tester,
        session: session(status: ChatStatus.assigned, handledBy: agent),
        fallbackTitle: kFallback,
      );
      expect(announcement(tester), "You're now chatting with Ada.");

      // And back to the NEW fallback, not the old one.
      await pump(
        tester,
        session: session(status: ChatStatus.waitingForAgent),
        fallbackTitle: kFallback,
      );
      expect(announcement(tester), "You're now chatting with $kFallback.");
    });
  });
}
